File: apple/Chuchotage/Network/SponsoredTrialClient.swift
import Foundation

final class SponsoredTrialClient: @unchecked Sendable {
    private let urlSession: URLSession
    private let clientSecretURL: URL
    private let retryDelaysNanoseconds: [UInt64]

    init(
        urlSession: URLSession = .shared,
        clientSecretURL: URL = URL(string: "https://www.chuchotage.ai/api/trial/realtime-translation-client-secret")!,
        retryDelaysNanoseconds: [UInt64] = [
            250_000_000,
            500_000_000,
            1_000_000_000,
            2_000_000_000,
        ]
    ) {
        self.urlSession = urlSession
        self.clientSecretURL = clientSecretURL
        self.retryDelaysNanoseconds = retryDelaysNanoseconds
    }

    func sessionBearerTokenFor(
        installID: String,
        targetLanguageCode: String = TranslationLanguages.defaultTargetLanguageCode,
        sourceTranscriptEnabled: Bool = false
    ) async throws -> RealtimeTranslationSessionToken {
        let clientSecret = try await createClientSecret(
            installID: installID,
            targetLanguageCode: TranslationLanguages.sanitizeOutputLanguageCode(targetLanguageCode),
            sourceTranscriptEnabled: sourceTranscriptEnabled
        )

        return RealtimeTranslationSessionToken(
            value: clientSecret,
            shouldSendSessionUpdate: false
        )
    }

    private func createClientSecret(
        installID: String,
        targetLanguageCode: String,
        sourceTranscriptEnabled: Bool
    ) async throws -> String {
        var request = URLRequest(url: clientSecretURL)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: [
            "installation_id": installID,
            "target_language": targetLanguageCode,
            "source_transcript_enabled": sourceTranscriptEnabled,
        ])

        return try await executeClientSecretRequestWithRetry(request)
    }

    private func executeClientSecretRequestWithRetry(_ request: URLRequest) async throws -> String {
        for attempt in retryDelaysNanoseconds.indices {
            do {
                let (data, response) = try await urlSession.data(for: request)
                guard let httpResponse = response as? HTTPURLResponse else {
                    throw SponsoredTrialClientError.requestFailed(
                        "Chuchotage returned an invalid response."
                    )
                }

                let text = String(data: data, encoding: .utf8) ?? ""
                let payload = parseJsonObject(data)

                switch httpResponse.statusCode {
                case 200..<300:
                    if let value = payload?["value"] as? String, !value.isEmpty {
                        return value
                    }
                    throw SponsoredTrialClientError.requestFailed(
                        "Chuchotage did not return a trial translation token."
                    )

                case 429:
                    throw SponsoredTrialClientError.requestFailed(
                        message(from: payload)
                            ?? "Sponsored trial limit reached. Sign in with ChatGPT or use an API key to continue."
                    )

                case 403:
                    throw SponsoredTrialClientError.requestFailed(
                        message(from: payload)
                            ?? "Sponsored trial is not available right now."
                    )

                case 503:
                    throw SponsoredTrialClientError.requestFailed(
                        "Sponsored trial is not available right now. Sign in with ChatGPT or use an API key to continue."
                    )

                case 500...599:
                    if attempt < retryDelaysNanoseconds.index(before: retryDelaysNanoseconds.endIndex) {
                        try await Task.sleep(nanoseconds: retryDelaysNanoseconds[attempt])
                        continue
                    }
                    throw SponsoredTrialClientError.requestFailed(
                        "Could not start sponsored trial. Check the phone's network and try again."
                    )

                default:
                    throw SponsoredTrialClientError.requestFailed(
                        errorMessage(from: payload, fallbackText: text)
                    )
                }
            } catch let error as SponsoredTrialClientError {
                throw error
            } catch let error as URLError {
                if isRetryableNetworkError(error),
                   attempt < retryDelaysNanoseconds.index(before: retryDelaysNanoseconds.endIndex) {
                    try await Task.sleep(nanoseconds: retryDelaysNanoseconds[attempt])
                    continue
                }
                throw SponsoredTrialClientError.network(networkMessage(for: error))
            }
        }

        throw SponsoredTrialClientError.requestFailed("Sponsored trial request failed.")
    }

    private func parseJsonObject(_ data: Data) -> [String: Any]? {
        try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    }

    private func message(from payload: [String: Any]?) -> String? {
        if let message = payload?["message"] as? String, !message.isEmpty {
            return message
        }
        return nil
    }

    private func errorMessage(from payload: [String: Any]?, fallbackText: String) -> String {
        message(from: payload)
            ?? (payload?["error"] as? String)
            ?? (fallbackText.isEmpty ? "Sponsored trial request failed." : fallbackText)
    }

    private func isRetryableNetworkError(_ error: URLError) -> Bool {
        switch error.code {
        case .cannotFindHost,
             .dnsLookupFailed,
             .timedOut,
             .cannotConnectToHost,
             .networkConnectionLost,
             .notConnectedToInternet:
            return true
        default:
            return false
        }
    }

    private func networkMessage(for error: URLError) -> String {
        switch error.code {
        case .cannotFindHost, .dnsLookupFailed:
            return L10n.string(
                "network.resolveChuchotage",
                defaultValue: "Could not resolve chuchotage.ai. Check the phone's VPN, private DNS, or network and try again."
            )
        default:
            return L10n.string(
                "network.reachChuchotage",
                defaultValue: "Could not reach chuchotage.ai. Check the phone's network and try again."
            )
        }
    }
}

enum SponsoredTrialClientError: LocalizedError, Sendable {
    case requestFailed(String)
    case network(String)

    var errorDescription: String? {
        switch self {
        case .requestFailed(let message), .network(let message):
            return message
        }
    }
}

File: app/src/main/java/com/andreabertoncini/chuchotage/network/SponsoredTrialClient.kt
package com.andreabertoncini.chuchotage.network

import com.andreabertoncini.chuchotage.settings.TranslationLanguages
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException
import java.net.SocketTimeoutException
import java.net.UnknownHostException

class SponsoredTrialClient(
    private val okHttpClient: OkHttpClient = OkHttpClient(),
    private val clientSecretUrl: String = CLIENT_SECRET_URL,
    private val retryDelaysMs: LongArray = CLIENT_SECRET_RETRY_DELAYS_MS,
) {
    suspend fun sessionBearerTokenFor(
        installId: String,
        targetLanguageCode: String = TranslationLanguages.DEFAULT_TARGET_LANGUAGE_CODE,
        sourceTranscriptEnabled: Boolean = false,
    ): RealtimeTranslationSessionToken {
        val clientSecret = createClientSecret(
            installId = installId,
            targetLanguageCode = TranslationLanguages.sanitizeOutputLanguageCode(targetLanguageCode),
            sourceTranscriptEnabled = sourceTranscriptEnabled,
        )
        return RealtimeTranslationSessionToken(
            value = clientSecret,
            shouldSendSessionUpdate = false,
            credentialKind = OpenAiCredentialKind.SPONSORED_TRIAL,
        )
    }

    private suspend fun createClientSecret(
        installId: String,
        targetLanguageCode: String,
        sourceTranscriptEnabled: Boolean,
    ): String = withContext(Dispatchers.IO) {
        val body = JSONObject()
            .put("installation_id", installId)
            .put("target_language", targetLanguageCode)
            .put("source_transcript_enabled", sourceTranscriptEnabled)
            .toString()
            .toRequestBody(JSON_MEDIA_TYPE)
        val request = Request.Builder()
            .url(clientSecretUrl)
            .post(body)
            .build()

        executeClientSecretRequestWithRetry(request)
    }

    private suspend fun executeClientSecretRequestWithRetry(request: Request): String {
        var lastNetworkError: IOException? = null

        retryDelaysMs.forEachIndexed { attempt, delayMs ->
            try {
                okHttpClient.newCall(request).execute().use { response ->
                    val text = response.body?.string().orEmpty()
                    val payload = parseJsonObject(text)

                    when {
                        response.isSuccessful -> {
                            return payload
                                ?.optString("value")
                                ?.takeIf { it.isNotBlank() }
                                ?: error("Chuchotage did not return a trial translation token.")
                        }
                        response.code == 429 -> throw SponsoredTrialException(
                            messageFrom(payload)
                                ?: "Chuchotage translation access is busy right now. Try again shortly.",
                        )
                        response.code == 403 -> throw SponsoredTrialException(
                            messageFrom(payload)
                                ?: "Chuchotage translation access is not available right now.",
                        )
                        response.code == 503 -> throw SponsoredTrialException(
                            "Chuchotage translation access is not available right now. Try again shortly.",
                        )
                        response.code in 500..599 -> {
                            if (attempt < retryDelaysMs.lastIndex) {
                                delay(delayMs)
                            } else {
                                throw SponsoredTrialException(
                                    "Could not start Chuchotage translation access. Check the phone's network and try again.",
                                )
                            }
                        }
                        else -> throw SponsoredTrialException(errorMessageFrom(payload, text))
                    }
                }
            } catch (exception: IOException) {
                lastNetworkError = exception
                if (!exception.isRetryableNetworkError() || attempt == retryDelaysMs.lastIndex) {
                    throw SponsoredTrialException(exception.toNetworkMessage(), exception)
                }
                delay(delayMs)
            }
        }

        val exception = lastNetworkError ?: IOException("Network request failed.")
        throw SponsoredTrialException(exception.toNetworkMessage(), exception)
    }

    private fun parseJsonObject(text: String): JSONObject? {
        return runCatching { JSONObject(text) }.getOrNull()
    }

    private fun errorMessageFrom(payload: JSONObject?, text: String): String {
        return messageFrom(payload)
            ?: payload?.optString("error")?.takeIf { it.isNotBlank() }
            ?: text.takeIf { it.isNotBlank() }
            ?: "Sponsored trial request failed."
    }

    private fun messageFrom(payload: JSONObject?): String? {
        return payload?.optString("message")?.takeIf { it.isNotBlank() }
    }

    private fun IOException.isRetryableNetworkError(): Boolean {
        return this is UnknownHostException || this is SocketTimeoutException
    }

    private fun IOException.toNetworkMessage(): String {
        return if (this is UnknownHostException) {
            "Could not resolve chuchotage.ai. Check the phone's VPN, private DNS, or network and try again."
        } else {
            "Could not reach chuchotage.ai. Check the phone's network and try again."
        }
    }

    class SponsoredTrialException(
        message: String,
        cause: Throwable? = null,
    ) : IllegalStateException(message, cause)

    companion object {
        private const val CLIENT_SECRET_URL =
            "https://www.chuchotage.ai/api/trial/realtime-translation-client-secret"
        private val CLIENT_SECRET_RETRY_DELAYS_MS = longArrayOf(250, 500, 1_000, 2_000)
        private val JSON_MEDIA_TYPE = "application/json".toMediaType()
    }
}
